File: src/main/java/org/elmlang/intellijplugin/features/structure/ElmFileTreeElement.java
package org.elmlang.intellijplugin.features.structure;

import com.intellij.ide.structureView.StructureViewTreeElement;
import com.intellij.ide.structureView.impl.common.PsiTreeElementBase;
import org.elmlang.intellijplugin.psi.ElmFile;
import org.elmlang.intellijplugin.psi.ElmValueDeclarationBase;
import org.elmlang.intellijplugin.utils.StreamUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collection;
import java.util.Comparator;
import java.util.stream.Collectors;

class ElmFileTreeElement extends PsiTreeElementBase<ElmFile> {
    @NotNull
    private ElmFile element;

    ElmFileTreeElement(@NotNull ElmFile psiElement) {
        super(psiElement);
        element = psiElement;
    }

    @Nullable
    public String getPresentableText() {
        return element.getName();
    }

    @NotNull
    @Override
    public Collection<StructureViewTreeElement> getChildrenBase() {
        return StreamUtil.concatAll(
                element.getValueDeclarations().map((ElmValueDeclarationBase psiElement) -> new ElmValueDeclarationTreeElement(psiElement)),
                element.getTypeAliasDeclarations().map(ElmTypeAliasTreeElement::new),
                element.getTypeDeclarations().map(ElmUnionTypeTreeElement::new)
        )
                .filter(a -> a.getElement() != null)
                .sorted(Comparator.comparingInt(a -> a.getElement().getTextOffset()))
                .collect(Collectors.toList());
    }
}

File: src/main/java/org/elmlang/intellijplugin/features/structure/ElmFileTreeElement.kt
package org.elmlang.intellijplugin.features.structure

import com.intellij.ide.structureView.StructureViewTreeElement
import com.intellij.ide.structureView.impl.common.PsiTreeElementBase
import org.elmlang.intellijplugin.psi.ElmFile
import java.util.stream.Collectors

class ElmFileTreeElement(element: ElmFile): PsiTreeElementBase<ElmFile>(element) {
    override fun getPresentableText(): String? {
        return element?.name
    }

    override fun getChildrenBase(): Collection<StructureViewTreeElement> {
        val file = element ?: return emptyList()

        val valueNodes = file.getValueDeclarations().map { ElmValueDeclarationTreeElement(it) }.collect(Collectors.toList())
        val typeAliasNodes = file.getTypeAliasDeclarations().map { ElmTypeAliasTreeElement(it) }.collect(Collectors.toList())
        val unionTypeNodes = file.getTypeDeclarations().map { ElmUnionTypeTreeElement(it) }.collect(Collectors.toList())

        return listOf(valueNodes, typeAliasNodes, unionTypeNodes)
                .flatten()
                .sortedBy { it.element?.textOffset }
    }
}